provider/openstack/firewaller.go

```diff
@@ -718,18 +719,32 @@ func (c *neutronFirewaller) openPortsInGroup(nameRegExp string, portRanges []net
 	return nil
 }
 
-// ruleMatchesPortRange checks if supplied nova security group rule matches the port range
-func ruleMatchesPortRange(rule neutron.SecurityGroupRuleV2, portRange network.PortRange) bool {
-	if rule.IPProtocol == nil || *rule.PortRangeMax == 0 || *rule.PortRangeMin == 0 {
+// secGroupMatchesIngressRule checks if supplied nova security group rule matches the ingress rule
+func secGroupMatchesIngressRule(secGroupRule neutron.SecurityGroupRuleV2, rule network.IngressRule) bool {
+	if secGroupRule.IPProtocol == nil || *secGroupRule.PortRangeMax == 0 || *secGroupRule.PortRangeMin == 0 {
 		return false
 	}
-	return *rule.IPProtocol == portRange.Protocol &&
-		*rule.PortRangeMin == portRange.FromPort &&
-		*rule.PortRangeMax == portRange.ToPort
+	portsMatch := *secGroupRule.IPProtocol == rule.Protocol &&
+		*secGroupRule.PortRangeMin == rule.FromPort &&
+		*secGroupRule.PortRangeMax == rule.ToPort
+	if !portsMatch {
+		return false
+	}
+	// The ports match, so if the security group RemoteIPPrefix matches *any* of the
+	// rule's source ranges, then that's a match.
+	if len(rule.SourceRanges) == 0 {
+		return secGroupRule.RemoteIPPrefix == "" || secGroupRule.RemoteIPPrefix == "0.0.0.0/0"
+	}
+	for _, r := range rule.SourceRanges {
```

Comment:
I'm pretty sure we should be sorting both sides, so we don't end up saying:
192.168.0.0/24, 25.25.25.25/24
!= 
25.25.25.25/24, 192.168.0.0/24

Or put them into a "strings.Set()" and compare them, etc. (Put them into a set, pull them out as we find them in the other one, assert the len of the set is 0.)